### ISSUE TITLE
Don't automatically register client modules anymore; make client registration explicit 

### ISSUE THREAD

**Issue by NobleDraconian** [evaera/Cmdr#101]
Currently, it is not possible for a game's client-sided systems to register their own commands, as Cmdr only supports command registering on the server.

This has an annoying side affect, if a client script wants to register its own command modules that are logically parented to it, the client command modules instead have to all be grouped together in a single folder on the server, and some arbitrary script has to be written to register the client sided commands. This is not ideal, as logically parenting command modules to the localscript that's registering them is more organized.

E.g.
```
SomeClientScript
   DebugCommands
      SomeClientCommand.lua
      SomeOtherClientCommand.lua
   init.lua
```

By allowing commands to be registered on the client, developers would be free to organize their file structure as they see fit on the client, and a game's client systems can worry about registering their own commands - instead of some arbitrary script on the server worrying about commands that should only exist on the client.

**Comment by LoganDark** [comment]
> To work around this

To work around what?

Are you working around commands needing to be registered on the server...?

Is this a separate issue?

**Comment by Reshiram110** [comment]
> To work around what?
Commands only being able to be registered on the server. Client-sided game systems cannot register their own commands, which makes the resulting file-structure disorganized as all of the game's commands have to be grouped under a single folder.

**Comment by LoganDark** [comment]
> Client-sided game systems cannot register their own commands, which makes the resulting file-structure disorganized

These are two separate issues. Make a new issue about nested folder structures (or nested command scripts, or however you want to phrase it).

**Comment by Reshiram110** [comment]
> Make a new issue about nested folder structures (or nested command scripts, or however you want to phrase it).

That's not the issue that I am facing, I am stating that client-sided systems cannot register their own commands, which means I cannot parent command modules to them. I'm not having any issues with nested command file structures.

E.g.
```
SomeClientScript
   DebugCommands
      SomeClientCommand.lua
      SomeOtherClientCommand.lua
   init.lua
```

As you can see in this example, the client script would want to register its own debug commands which are logically parented to it in the file structure. But since registering commands on the client is not supported by Cmdr, this file structure is not possible.

**Comment by LoganDark** [comment]
My bad, you should have provided an example in the post.

Your proposal is for commands that run on the client, right? That's a valid feature request, I don't see any problem with it. I just think you could make the post a little more clear.

**Comment by Reshiram110** [comment]
I'll update the first post to have more concise wording.

**Comment by evaera** [comment]
I think you should be able to run `CmdrClient.Registry:RegisterCommandsIn`. The only thing that's stopping you is the documentation.

**Comment by LoganDark** [comment]
@evaera additionally: reading the source code is documentation :)

**Comment by evaera** [comment]
I want to completely remove the automatic client registration when registered on the server, and instead let developers register both independently. Cmdr modules around causes problems in more way than one.

This is a big breaking change though, so I've been putting it off for a big revamp. I don't really know when I'll have time to do that revamp, but it's something I want to do.

**Comment by LoganDark** [comment]
I would second guess that idea. I think you should keep the ability to auto-register commands on both sides, but allow developers to register commands on the server side or client side only. Registering something on both sides is still a major use case for Cmdr but I would definitely appreciate some flexibility if you only want a command on one side.

Server side commands wouldn't be autocompleted or visible to the client, useful for like an easter egg / secret debug mode or whatever. Client side commands wouldn't touch the server and would be executed immediately by the client. Right?
